Foundation/List/DList.hs

```diff
@@ -0,0 +1,89 @@
+-- |
+-- Module      : Foundation.List.DList
+-- License     : BSD-style
+-- Maintainer  : Nicolas Di Prima <nicolas@primetype.co.uk>
+-- Stability   : statble
+-- Portability : portable
+--
+-- Data structure for optimised operations (append, cons, snoc) on list
+--
+module Foundation.List.DList
+    ( DList
+    ) where
+
+import Foundation.Internal.Base
+import Foundation.Collection
+import Foundation.Class.Bifunctor
+
+newtype DList a = DList { unDList :: [a] -> [a] }
+  deriving (Typeable)
+
+instance Eq a => Eq (DList a) where
+    (==) dl1 dl2 = (==) (toList dl1) (toList dl2)
+
+instance Ord a => Ord (DList a) where
+    compare dl1 dl2 = compare (toList dl1) (toList dl2)
+
+instance Show a => Show (DList a) where
+    show = show . toList
+
+instance IsList (DList a) where
+    type Item (DList a) = a
+    fromList = DList . (<>)
+    toList = flip unDList []
+
+instance Monoid (DList a) where
+    mempty = DList id
+    mappend dl1 dl2 = DList $ unDList dl1 . unDList dl2
+
+instance Functor DList where
+    fmap f = foldr (cons . f) mempty
+
+instance Applicative DList where
+    pure = singleton
+    (<*>) m1 m2 = m1 >>= \x1 -> m2 >>= \x2 -> return (x1 x2)
+
+instance Monad DList where
+    (>>=) m k = foldr (mappend . k) mempty m
+    return = singleton
+    fail _ = mempty
```

Comment:
we don't need fail, it's going to be in MonadFail, and DList won't have the instance anyway